src/main/scala/rocket/DCache.scala

```diff
@@ -449,19 +459,24 @@ class DCacheModule(outer: DCache) extends HellaCacheModule(outer) {
 
   // grant
   val (d_first, d_last, d_done, d_address_inc) = edge.addr_inc(tl_out.d)
-  val grantIsCached = {
-    val res = tl_out.d.bits.opcode.isOneOf(Grant, GrantData)
+  val (d_opc, grantIsUncached, grantIsUncachedData) = {
+    val uncachedGrantOpcodesSansData = Seq(AccessAck, HintAck)
+    val uncachedGrantOpcodesWithData = Seq(AccessAckData)
+    val uncachedGrantOpcodes = uncachedGrantOpcodesWithData ++ uncachedGrantOpcodesSansData
+    val whole_opc = tl_out.d.bits.opcode
     if (usingDataScratchpad) {
-      assert(!(tl_out.d.valid && res))
-      false.B
+      assert(!tl_out.d.valid || whole_opc.isOneOf(uncachedGrantOpcodes))
+      // the only valid TL-D messages are uncached, so we can do some pruning
+      val opc = whole_opc(uncachedGrantOpcodes.map(_.getWidth).max - 1, 0)
+      val data = DecodeLogic(opc, uncachedGrantOpcodesWithData, uncachedGrantOpcodesSansData)
+      (opc, true.B, data)
```

Comment:
Seems like all of this should get sucked into some kind of helper method on `TLMessages`?

Reply:
Per in-person discussion, postponing until TL message-ID mapping is revisited and associated code is refactored.